### ISSUE TITLE
nbgrader as a service?

### ISSUE THREAD

**Issue by geoff-va** [jupyter/nbgrader#949]
Is it possible to user nbgrader as a service instead of requiring the entire infrastructure, eg, filesystem layout, database, etc?

The workflow I'm looking for is something more akin to:
* Generate instructor source file via a local instance of nbgrader/jupyter
* Generating a student version via `NbGraderAPI.assign`
* Grading a single students notebook via `NbGraderAPI.autograde`
* return auto-graded results in a `json` string.

Essentially, I'm not interested in the database component - just an interface to say:
* here's an instructor notebook
* here's is student notebook
* here are the results of autograding that student notebook against the instructor notebook

Maybe this is possible with the current API and I just haven't determined how to make that work? But so far, seems everything requires a backend of some kind.

**Comment by jhamrick** [comment]
Unfortunately, this functionality is not currently supported and it would require a major redesign of nbgrader to add it 😞 

It is currently possible to generate the assignments without the database (by passing the `--no-db` flag or setting `Assign.no_database = True` and `SaveCells.enabled = False`), but this still requires the filesystem layout.

A similar functionality does not exist for autograding. Much of the heavy lifting that the autograder does is to do things like ensure the student hasn't changed test cells (and overwrite the contents of those cells with what's stored in the database if they have changed). I could have the autograder bypass the database but then you would lose this functionality (and importantly there would be no guarantees against students having done something like change the number of points a cell is worth). A significant redesign of nbgrader could optionally pull this information from a reference notebook rather than from the database, but this would unfortunately be something that I think is outside the current scope of the project.

If you are interested in having the autograder bypass the database (with the caveats above) I can certainly add that feature, and then add a new method to the api (something like `autograde_json`) which returns the json results of grading. However, it would unfortunately still depend on the filesystem layout.

**Comment by geoff-va** [comment]
No, I don't think having the autograder bypass those checks is worthwhile. Accurate results are more important than convenient results :) Thanks though!
